### PR TITLE
Fixes #2118 Add a Publications Content Type to QS2

### LINKED ISSUE #2118: Add a Publications Content Type to QS2
## Motivation
Multiple sites need a Publication feature, ie. Law School, Sbar, Tree Lab, Pharmacy.

### Is your feature request related to a problem? Please describe.
No.

## Proposed Resolution
Create a Publications Content Type that can encompass MLS style, IEEE style, AMA style and Bluebook style at the very least. Possibly include Chicago style, and APA styles as well as others. Each style should be created as a different View.

### Describe the solution you'd like
The Content Type should have fields for the follow:
Author - Text or Entity Reference (this should be large enough to accommodate at least 3 author's names)
Title of Container- Text (this should also be fairly large and allow for italics, underlining etc.)
Title of Piece - Text (this should also be fairly large and allow for italics, underlining etc.)
Other contributors - Text (should be around 200 characters)
Version - Text 
Number - Possibly Text (will contain volume or number)
Publisher - Text (approximately 100 characters)
Publication Date - May have to be Text field, as the style of date varies
Publication Type - Text
Location - Text
Page - Could be Integer, depending on formatting needs
Links - There needs to be the option to have several: used for DOI number link, BibTex, Google Scholar, PubMed or other links to online sources.
Links to downloadable material, PDF

Search features:
Search by Publication Year/Date
Search by Publication Type (ie. Book, journal, video)
Search by Author
Search by Keyword in Title of Piece

Link to U of Az "Citing Your Sources" for further info on guidelines for styles: https://libguides.library.arizona.edu/citing

This list is PRELIMINARY. I am hoping to get feedback on the above list for inclusions or removals.

CSL rendering: https://citationstyles.org/

### Describe alternatives you've considered
Using a table, grid or a custom feature for each site. None of these are the best option

### Roles and Permissions considerations
A clear and concise description of how each of the following roles would be impacted by this change:
- Anonymous user - use reference material on the website
- Authenticated user - use reference material on the website
- Content editor - easier to enter and maintain Publications content
- Content administrator - easier to enter and maintain Publications content
- Administrator - Clients will be much happier to have a well maintained Content type so the Admins can sleep better.

## Additional context
None at this time.

### PR DESCRIPTION
**Pull request by tadean**
## Description

This PR adds a Pupblication content type to Quickstart. It contains a Publication content type, an Author entity that is used for representing publication authors (with optional referential linking to a Quickstart Person). It presently contains styles for MLA, IEEE, AMA, Bluebook Law Review, APA, and Chicago styles. It uses [Citation Style Langauge](https://citationstyles.org/) for bibliographic rendering, via a PHP CSL 1.0.1 processor ([seboettg/citeproc-php](https://github.com/seboettg/citeproc-php))

- Additional views to be added
- Role permissions to be added

## Related issues

#2118 

## How to test

- Enable **Quickstart Publication (Experimental)**
- Create a flexible page that adds displays of the AZ Publications views to the content
- The AZ Publication views currently have one display per citation style standard
- Create publication content
- View the publication output on your view page for correctness
- Not all fields are used by every publication type and style standard. Depending on the type of publication and citation style some fields may not be displayed if the style does not use them.

### Arizona Quickstart (install profile, custom modules, custom theme)
- **Patch release changes**
   - [ ] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
   - [x] Adding experimental module
   - [ ] Update experimental module
- **Minor release changes**
   - [x] New feature
   - [ ] Breaking or visual change to existing behavior
   - [ ] Upgrade experimental module to stable
   - [ ] Enable existing module by default or database update
   - [ ] Non-critical brand change
   - [x] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- **Other or unknown**
   - [ ] Other or unknown

### Drupal core
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major or minor level update
- **Other or unknown**
   - [ ] Other or unknown

### Drupal contrib projects
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch or minor level update
   - [x] Add new module
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major level update
- **Other or unknown**
   - [ ] Other or unknown

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by tadean** [comment]
**Todo:**

- [ ] Fix phpstan entity warnings
- [ ] Add permissions (pending discussion)
- [ ] Potentially move stylesheets into configuration entities (pending discussion)
- [ ] Render additional links in citation
- [ ] Add additional views

**Comment by UAScotttJaxon** [comment]
Can we include PubMed ID in this module? College of Medicine/UAHS sites utilize PubMed ID. https://www.drupal.org/project/bibcite_pubmed

**Review by joshuasosa — COMMENTED** [review]
Are all these locale files necessary? It seems like AZQS is primarily focused on English and doesn't provide any other translations for anything else. Could citation-style-language/locales be reduced to mostly just EN to contain the scope?

**Comment by tadean** [comment]
> Are all these locale files necessary? It seems like AZQS is primarily focused on English and doesn't provide any other translations for anything else. Could citation-style-language/locales be reduced to mostly just EN to contain the scope?

I agree - I'm hoping to make a change that moves the locale files to the vendor folder (via a CSL metapackage) and moves the stylesheets to configuration entities so other stylesheets can be added to the site without needing to change the module.

**Comment by tadean** [comment]
> Can we include PubMed ID in this module? College of Medicine/UAHS sites utilize PubMed ID. https://www.drupal.org/project/bibcite_pubmed

We discussed adding something for this today and it seemed reasonable. The one caveat is that not all style standards have a a rule for displaying a pubmed (or for example, a DOI) link, so whether that ID is used would depend on the style standard.

**Comment by tadean** [comment]
Todo:
- [ ] Move stylesheets into configuration entities
- [ ] Investigate moving locales to `/vendor`
- [ ] Add additional views
- [ ] Investigate pubmed/DOI ID

**Comment by tadean** [comment]
followup research note: PubMed IDs appear as a custom variable in the National Library of Medicine style and several others based upon it. It is typically exposed to CSL processing as the variable `PMID`

DOI numbers are part of the CSL standard schema and use the identifier `DOI`.

**Comment by tadean** [comment]
- Removed CSL stylesheets as hardcoded module files in favor of a new Citation Style configuration entity type. This moves the stylesheets into configuration and makes it possible to add additional styles to a site without editing the module. Administration is at `/admin/structure/az_citation_style`
- Removed locale files from the module in favor of using `academicpuma/locales` and adding a patch to allow the CSL processor to locate and use them. This greatly reduces the size of the module but makes it more difficult to change the locale files if needed. However, the locale files change **much** less frequently than the stylesheets and I propose we can make any occasional modifications to the locale files that are needed via composer patches.

**Comment by tadean** [comment]
Updated the drupal views patch for the rendered content issue to be compatible with 9.5.x.

File: modules/custom/az_publication/az_publication.module [hunk 1]
```diff
@@ -0,0 +1,448 @@
...
+ */
+function az_publication_node_view(array &$build, EntityInterface $entity, EntityViewDisplayInterface $display, $view_mode) {
+
+  // Only if fields suggest a publication.
+  if ($entity instanceof NodeInterface &&
+      $entity instanceof FieldableEntityInterface &&
+      $entity->hasField('field_az_publication_type') &&
+      !empty($entity->field_az_publication_type->value)) {
+    if ($display->getComponent('az_publication_bibliography')) {
+      // Publicaiton fields to CSL schema mappings.
```
Comment (bberndt-uaz) [inline modules/custom/az_publication/az_publication.module:77]:
```suggestion
      // Publication fields to CSL schema mappings.
```

File: modules/custom/az_publication/az_publication.module [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+}
+
+/**
+ * Create a CSL schema date object with date-parts based on a entity.
+ *
+ * To send a a publication for CSL processing, dates need to be transformed
+ * into CSL format. This is an object with a date-parts property and the primary
+ * processing needing to be done is making sure the date-parts nested array has
+ * the number of components dictated by the date type, e.g. we do not want to
+ * send a date with three componens if there should be only a year component.
```
Comment (bberndt-uaz) [inline modules/custom/az_publication/az_publication.module:249]:
```suggestion
 * send a date with three components if there should be only a year component.
```

File: modules/custom/az_publication/az_publication.info.yml [hunk 3]
```diff
@@ -0,0 +1,31 @@
+name: Quickstart Publication (Experimental)
+description: 'Provides Publication content type'
+core_version_requirement: ^9
```
Comment (joeparsons) [inline modules/custom/az_publication/az_publication.info.yml:3]:
```suggestion
core_version_requirement: ^9 || ^10
```

**Review by joeparsons — APPROVED** [review]

**Review by camikazegreen — APPROVED** [review]

**Pull request merged**